src/main/java/com/splunk/kafka/connect/KafkaRecordTracker.java

```diff
@@ -22,21 +22,67 @@
 import org.apache.kafka.connect.sink.SinkRecord;
 
 import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicLong;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 final class KafkaRecordTracker {
     private static final Logger log = LoggerFactory.getLogger(SplunkSinkTask.class);
-    private Map<TopicPartition, TreeMap<Long, EventBatch>> all; // TopicPartition + Long offset represents the SinkRecord
-    private long total;
+    private ConcurrentHashMap<TopicPartition, TreeMap<Long, EventBatch>> all; // TopicPartition + Long offset represents the SinkRecord
+    private AtomicLong total;
     private ConcurrentLinkedQueue<EventBatch> failed;
+    private volatile Map<TopicPartition, OffsetAndMetadata> offsets;
 
     public KafkaRecordTracker() {
-        all = new HashMap<>();
+        all = new ConcurrentHashMap<>();
         failed = new ConcurrentLinkedQueue<>();
-        total = 0;
+        total = new AtomicLong();
+        offsets = new HashMap<>();
+    }
+
+    public void removeAckedEventBatches(final List<EventBatch> batches) {
+        for (final EventBatch batch: batches) {
+            //log.debug("Processing batch {}", batch.getUUID());
+            removeAckedEventBatch(batch);
+        }
+    }
+
+    public void removeAckedEventBatch(final EventBatch batch) {
+        final List<Event> events = batch.getEvents();
+        final Event event = events.get(0);
+        if (event.getTied() instanceof SinkRecord) {
+            final SinkRecord record = (SinkRecord) event.getTied();
+            TopicPartition tp = new TopicPartition(record.topic(), record.kafkaPartition());
+            //log.debug("Processing topic {} partition {}", record.topic(), record.kafkaPartition());
+            TreeMap<Long, EventBatch> tpRecords = all.get(tp);
+            if (tpRecords == null) {
+                log.error("KafkaRecordTracker removing a batch in an unknown partition {} {} {}", record.topic(), record.kafkaPartition(), record.kafkaOffset());
+                return;
+            }
+            long offset = -1;
+            Iterator<Map.Entry<Long, EventBatch>> iter = tpRecords.entrySet().iterator();
+            for (; iter.hasNext();) {
+                Map.Entry<Long, EventBatch> e = iter.next();
+                if (e.getValue().isCommitted()) {
+                    //log.debug("processing offset {}", e.getKey());
+                    offset = e.getKey();
+                    iter.remove();
+                    total.decrementAndGet();
+                } else {
+                    break;
+                }
+            }
+            if (offset >= 0) {
+                if (offsets.containsKey(tp)) {
+                    offsets.replace(tp, new OffsetAndMetadata(offset + 1));
```

Comment:
If we check the key exists, there is no differences between replace & put? I believe one `put` is enough.

Reply:
oh yeah, in this case only one put will be sufficient